src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal sealed class OpenSslPkcs12Reader : IDisposable
+    {
+        private readonly SafePkcs12Handle _pkcs12Handle;
+        private SafeEvpPkeyHandle _evpPkeyHandle;
+        private SafeX509Handle _x509Handle;
+        private SafeX509StackHandle _caStackHandle;
+
+        private OpenSslPkcs12Reader(SafePkcs12Handle pkcs12Handle)
+        {
+            _pkcs12Handle = pkcs12Handle;
+        }
+
+        public unsafe static OpenSslPkcs12Reader TryRead(byte[] data)
+        {
+            SafePkcs12Handle handle = Interop.libcrypto.OpenSslD2I(
+                (ptr, b, i) => Interop.libcrypto.d2i_PKCS12(ptr, b, i),
+                data,
+                checkHandle: false);
+
+            if (handle != null && !handle.IsInvalid)
+            {
+                return new OpenSslPkcs12Reader(handle);
+            }
+
+            return null;
+        }
+
+        public static OpenSslPkcs12Reader TryRead(SafeBioHandle fileBio)
+        {
+            SafePkcs12Handle p12 = Interop.libcrypto.d2i_PKCS12_bio(fileBio, IntPtr.Zero);
+
+            if (p12 != null && !p12.IsInvalid)
+            {
+                return new OpenSslPkcs12Reader(p12);
+            }
+
+            return null;
+        }
+
+        public void Dispose()
+        {
+            if (_caStackHandle != null)
+            {
+                _caStackHandle.Dispose();
+                _caStackHandle = null;
+            }
+
+            if (_x509Handle != null)
+            {
+                _x509Handle.Dispose();
+                _x509Handle = null;
+            }
+
+            if (_evpPkeyHandle != null)
+            {
+                _evpPkeyHandle.Dispose();
+                _evpPkeyHandle = null;
+            }
+
+            if (_pkcs12Handle != null)
+            {
+                _pkcs12Handle.Dispose();
+            }
+        }
+
+        public void Decrypt(string password)
+        {
+            bool parsed = Interop.libcrypto.PKCS12_parse(
+                _pkcs12Handle,
+                password,
+                out _evpPkeyHandle,
+                out _x509Handle,
+                out _caStackHandle);
+
+            if (!parsed)
+            {
+                throw Interop.libcrypto.CreateOpenSslCryptographicException();
+            }
+        }
```

Comment:
I can envision situations in which OpenSsl could emit `NULL` as the answer to any of those (single cert: don't allocate _caStackHandle; no private key, don't allocate _evpPkeyHandle; no private key, don't allocate _x509Handle either... move it into _caStack); so assuming that they just give us "some stuff" and we deal with it feels the most future-proofed.